src/server.cpp

```diff
@@ -1500,18 +1500,20 @@ void Server::SendMovement(u16 peer_id)
 	Send(&pkt);
 }
 
-void Server::SendPlayerHPOrDie(PlayerSAO *playersao)
+void Server::SendPlayerHPOrDie(PlayerSAO *playersao, s16 oldhp)
 {
 	if (!g_settings->getBool("enable_damage"))
 		return;
 
 	u16 peer_id   = playersao->getPeerID();
 	bool is_alive = playersao->getHP() > 0;
 
-	if (is_alive)
-		SendPlayerHP(peer_id);
-	else
+	// Make sure that the player only dies once.
+	if (playersao->getHP() < oldhp && !is_alive) {
```

Comment:
@kwolekr There's also currently a bug (tested in 0.4.14, but the cause is still there) where if you die, then disconnect without respawn (close the window) and re-connect, the death functions will be called twice, and two deathscreen packets will be sent.
Technically fixable with a one-line patch right now (SendPlayerHPOrDie -> SendPlayerHP, since the Deathscreen code is already called later anyway), but it's all linked to the same underlying issue you describe of SendPlayerHPOrDie being a kludge.
https://gist.github.com/20kdc/d984f7af1428d3445a3b14cd41506eb0